kotlinx-coroutines-core/wasm/src/internal/CopyOnWriteList.kt

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2016-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.internal
+
+import kotlinx.atomicfu.*
+
+// Copy of Native implementation
+
+@Suppress("UNCHECKED_CAST")
+internal class CopyOnWriteList<E> : AbstractMutableList<E>() {
+
+    private val _array = atomic(arrayOfNulls<Any?>(0))
```

Comment:
Is this implementation supposed to be used in multithreaded environments?

If not, JS has COW implementation without atomics